x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -0,0 +1,298 @@
+            .put("dynamic_templates", templates)
+            .map();
+    }
+
+    private void prepareAndCreateEnrichIndex(final String policyName, final EnrichPolicy policy,
+                                             final ActionListener<EnrichPolicyResult> listener) {
+        long nowTimestamp = nowSupplier.getAsLong();
+        String enrichIndexName = getEnrichIndexBase(policyName) + "-" + nowTimestamp;
+        // TODO: Settings for localizing enrich indices to nodes that are ingest+data only
+        Settings enrichIndexSettings = Settings.EMPTY;
+        Map<String, Object> enrichMapping = resolveEnrichMapping(policy);
+        CreateIndexRequest createEnrichIndexRequest = new CreateIndexRequest(enrichIndexName, enrichIndexSettings);
+        createEnrichIndexRequest.mapping("_doc", enrichMapping);
+        logger.debug("Policy [{}]: Creating new enrich index [{}]", policyName, enrichIndexName);
+        client.admin().indices().create(createEnrichIndexRequest, new ActionListener<CreateIndexResponse>() {
+            @Override
+            public void onResponse(CreateIndexResponse createIndexResponse) {
+                transferDataToEnrichIndex(enrichIndexName, policyName, policy, listener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private void transferDataToEnrichIndex(final String destinationIndexName, final String policyName, final EnrichPolicy policy,
+                                           final ActionListener<EnrichPolicyResult> listener) {
+        logger.debug("Policy [{}]: Transferring source data to new enrich index [{}]", policyName, destinationIndexName);
+        // Filter down the source fields to just the ones required by the policy
+        final Set<String> retainFields = new HashSet<>();
+        retainFields.add(policy.getEnrichKey());
+        retainFields.addAll(policy.getEnrichValues());
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.fetchSource(retainFields.toArray(new String[0]), new String[0]);
+        if (policy.getQuery() != null) {
+            searchSourceBuilder.query(QueryBuilders.wrapperQuery(policy.getQuery().getQuery()));
+        } else {
+            searchSourceBuilder.query(QueryBuilders.matchAllQuery());
+        }
+        ReindexRequest reindexRequest = new ReindexRequest()
+            .setDestIndex(destinationIndexName)
+            .setSourceIndices(policy.getIndexPattern());
+        reindexRequest.getSearchRequest().source(searchSourceBuilder);
+        client.execute(ReindexAction.INSTANCE, reindexRequest, new ActionListener<BulkByScrollResponse>() {
+            @Override
+            public void onResponse(BulkByScrollResponse bulkByScrollResponse) {
+                // Do we want to fail the request if there were failures during the reindex process?
```

Comment:
I am not familiar with this API... does this get called once after the whole re-index is complete, or per scroll response ?

Reply:
this gets invoked after the whole reindex has been completed.